### PR TITLE
Write permission for Users to global persist dir

### PR DESCRIPTION
**Pull request by Retia-Adolf**
lukesampson/scoop-extras#1124

### CODE REVIEW

**Comment by rasa** [comment]
files have no lines containing trailing whitespace 149ms
30      RuntimeException: The following 1 lines contain trailing whitespace: 
31      File: C:\projects\scoop\lib\install.ps1, Line: 58

**Comment by rasa** [comment]
files have lines containing trailing whitespace.
RuntimeException: The following 1 lines contain trailing whitespace: 
File: C:\projects\scoop\lib\install.ps1, Line: 58

**Comment by Retia-Adolf** [comment]
??! I had thought these user/group names are all the same. Never see differences on a Chinese PC etc.

I think I need to find a way to get the actual `Users` group name, else it will continue to fail on other European PCs...

**Comment by r15ch13** [comment]
Found a solution https://stackoverflow.com/questions/40587096/set-output-language-of-get-acl#40588213
`$user = New-Object System.Security.Principal.SecurityIdentifier 'S-1-5-32-545'`

**Comment by r15ch13** [comment]
Fixed it

Call:
```powershell
    # persist data
    persist_data $manifest $original_dir $persist_dir
    persist_permission $manifest $global
```

Function:
```powershell
# check whether write permission for Users usergroup is set to global persist dir, if not then set
function persist_permission($manifest, $global) {
    if (!$manifest.persist -and !$global) {
        return
    }
    $path = persistdir $null $global
    $sid = New-Object System.Security.Principal.SecurityIdentifier 'S-1-5-32-545'
    $user = ($sid.Translate([System.Security.Principal.NTAccount])).Value
    $acl = Get-Acl -Path $path

    $target_rule = New-Object System.Security.AccessControl.FileSystemAccessRule($user, 'Write', 'ObjectInherit', 'none', 'Allow')

    $is_set = $true
    ForEach ($exist_rule in $acl.Access) {
        $is_set = ($exist_rule -match $target_rule) -and $is_set
    }

    if (!$is_set) {
        $acl.SetAccessRule($target_rule)
        $acl | Set-Acl -Path $path
    }
}
```

**Comment by r15ch13** [comment]
Fixed the code (See above)

File: lib/install.ps1 [hunk 1]
```diff
@@ -1220,3 +1222,31 @@ function persist_data($manifest, $original_dir, $persist_dir) {
         }
     }
 }
+
+#check whether write permission for Users usergroup is set to global persist dir, if not then set
+function persist_permission($persist, $global) {
+    if ($persist -and $global) {
+        $path = "$(basedir $global)\persist"
+        $user = -join ([System.Environment]::MachineName, '\',
+            (([System.Security.Principal.SecurityIdentifier]'S-1-5-32-545').Translate([System.Security.Principal.NTAccount])).Value)
```
Comment (r15ch13) [inline lib/install.ps1]: This would result in: `R15CH13-PC\VORDEFINIERT\Benutzer`

Comment (r15ch13) [reply lib/install.ps1]: The important part is "testing"
I just installed and uninstalled notepad++ with these changes about 30 times now 😁

**Comment by Retia-Adolf** [comment]
@r15ch13 Checking permission seems no problem for me:
```powershell
❯ $path = $PWD
❯ $path

Path
----
C:\Users\Retia\Git\Retia-Adolf

❯ $user = [System.Security.Principal.SecurityIdentifier]'S-1-5-32-545'
❯ $Rights = "Write"
❯ $InheritSettings = "ObjectInherit"
❯ $PropogationSettings = "none"
❯ $RuleType = "Allow"
❯ $acl = Get-Acl -Path $PWD
❯ $perm = $user, $Rights, $InheritSettings, $PropogationSettings, $RuleType
❯ $targetRule = New-Object -TypeName System.Security.AccessControl.FileSystemAccessRule -ArgumentList $perm
❯ $isSet = $true
❯ ForEach ($existRule in $acl) {
>>     $isSet = ($existRule -match $targetRule) -and $isSet
>> }
❯ $isSet
False
```
So `!($isSet)` will equal `$true`, then `if (!($isSet))` should perform.

**Comment by r15ch13** [comment]
This can't work, because it should be `ForEach ($existRule in $acl.Access)` instead of `ForEach ($existRule in $acl)`.
Comparing the rules with `-match` also doesn't work. Add `debug ($existRule -match $targetRule)` to the Loop and you will see that it always outputs `True` which it shouldn't.

**Comment by Retia-Adolf** [comment]
Thanks for explaining

**Comment by Ash258** [comment]
```
Downloads $ scoop install sysinternals
Installing 'sysinternals' (July.5.2018) [64bit]
SysinternalsSuite.zip (24.0 MB) [=============================================================================] 100%
Checking hash of SysinternalsSuite.zip ... ok.
Extracting SysinternalsSuite.zip ... done.
Linking S:\Scoop\apps\sysinternals\current => S:\Scoop\apps\sysinternals\July.5.2018
Creating shim for 'accesschk'.
...Lots of shims...
Creating shortcut for SysInternals/ProcessExplorer - Enhanced Task Manager (procexp64.exe)
...Lots of shortcuts...
Set-Acl : Attempted to perform an unauthorized operation.
At S:\Scoop\apps\scoop\current\lib\install.ps1:1235 char:12
+     $acl | Set-Acl -Path $path
+            ~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : PermissionDenied: (S:\Scoop\persist\:String) [Set-Acl], UnauthorizedAccessException
    + FullyQualifiedErrorId : System.UnauthorizedAccessException,Microsoft.PowerShell.Commands.SetAclCommand
```

**Comment by r15ch13** [comment]
@Ash258 oh shit! Fixed with https://github.com/lukesampson/scoop/commit/eb7b7cbf4f30e4122762856723155f3c1e980d1b

**Comment by Ash258** [comment]
@r15ch13 Everything works fine now 😍

**Comment by Retia-Adolf** [comment]
Oops, I didn't consider the case that standard user installs app in a custom global app's directory.

**Pull request merged**